base/inference.jl

```diff
@@ -2607,21 +2607,12 @@ function tmerge(@nospecialize(typea), @nospecialize(typeb))
     if !(isa(typea,Type) || isa(typea,TypeVar)) || !(isa(typeb,Type) || isa(typeb,TypeVar))
         return Any
     end
-    if (typea <: Tuple) && (typeb <: Tuple)
-        if isa(typea, DataType) && isa(typeb, DataType) && length(typea.parameters) == length(typeb.parameters) && !isvatuple(typea) && !isvatuple(typeb)
-            return typejoin(typea, typeb)
-        end
-        if isa(typea, Union) || isa(typeb, Union) || (isa(typea,DataType) && length(typea.parameters)>3) ||
-            (isa(typeb,DataType) && length(typeb.parameters)>3)
-            # widen tuples faster (see #6704), but not too much, to make sure we can infer
-            # e.g. (t::Union{Tuple{Bool},Tuple{Bool,Int}})[1]
-            return Tuple
-        end
-    end
     u = Union{typea, typeb}
-    if unionlen(u) > MAX_TYPEUNION_LEN || type_too_complex(u, MAX_TYPE_DEPTH)
-        # don't let type unions get too big
-        # TODO: something smarter, like a common supertype
+    if unionlen(u) > MAX_TYPEUNION_LEN
+        u = typejoin(typea, typeb)
```

Comment:
Correct me if I'm wrong, but ensuring convergence requires monotonicity in the sense that `a <: a´ && b <: b´` implies `tmerge(a, b) <: tmerge(a´, b´)`, right? Then indeed we cannot let `typejoin` do the core work because of e.g. this:
```julia
julia> a´ = a = Ref{Int}
Ref{Int64}

julia> b = Ref{UInt}
Ref{UInt64}

julia> b´ = Ref{<:Integer}
Ref{#s1} where #s1<:Integer

julia> typejoin(a, b)
Ref

julia> typejoin(a, b´)
Ref{#s1} where #s1<:Integer

julia> a <: a´ && b <: b´
true

julia> typejoin(a, b) <: typejoin(a, b´)
false
```
But turns out we're already using `typejoin` in `tmerge` for (certain) tuples. And indeed (on master and 0.6):
```julia
julia> a´ = a = Tuple{Int32, Ref{Int}}
Tuple{Int32,Ref{Int64}}

julia> b = Tuple{Int64, Ref{UInt}}
Tuple{Int64,Ref{UInt64}}

julia> b´ = Tuple{Int64, Ref{<:Integer}}
Tuple{Int64,Ref{#s30} where #s30<:Integer}

julia> Core.Inference.tmerge(a, b)
Tuple{Signed,Ref}

julia> Core.Inference.tmerge(a, b´)
Tuple{Signed,Ref{#s30} where #s30<:Integer}

julia> a <: a´ && b <: b´
true

julia> Core.Inference.tmerge(a, b) <: Core.Inference.tmerge(a, b´)
false
```
Unless someone tells me I'm off track here and this is not an issue, my idea would be to first rethink `typejoin` and depending on the outcome either go ahead with this PR or move in the other direction and completely remove `typejoin` from `tmerge`.